### PR TITLE
original_publisher, publisher, and based_near Batch updates

### LINKED ISSUE #634: Batch metadata field updates
List of metadata fields to be changed in DigitalHub in batch if possible:

-All items with "Pediatric Neurology Briefs Publishers" in the Original Publisher field --> Add additional field to Digital Publisher with "Pediatric Neurology Briefs Publishers" and delete from Original Publisher
-All items in Science in Society Collection - delete "Science in Society" from Original Publisher
-All DigitalHub items - delete "Chicago, Illinois, United States" and delete "Chicago, IL, USA" from Location

It would be great if this could be done in next 2 months. Thanks!

### PR DESCRIPTION
**Pull request by sharpattack**
- closes #634

### CODE REVIEW

File: lib/batch/batch_metadata_updates_634.rb [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+#   - Northwestern University" --> change to "Galter Health Sciences Library"
+
+ghsl_fsm_nu_publisher = ["Galter Health Sciences Library, Feinberg School of Medicine, Northwestern University"]
+items_needing_publisher_updates = GenericFile.where("publisher" => ghsl_fsm_nu_publisher) + 
+                                  Collection.where("publisher" => ghsl_fsm_nu_publisher)
+
+items_needing_publisher_updates.each do |item|
+  item.publisher -= ghsl_fsm_nu_publisher
+  item.save
+end
```
Comment (fenekku) [inline lib/batch/batch_metadata_updates_634.rb]: You were saying you had issues with the script not updating records. How did you test that this works?

File: lib/batch/batch_metadata_updates_634.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+#   - All items with "Pediatric Neurology Briefs Publishers" in the Original 
+#   - Publisher field --> Add additional field to Digital Publisher with "Pediatric 
+#   - Neurology Briefs Publishers" and delete from Original Publisher
+
+pnb_publishers = "Pediatric Neurology Briefs Publishers"
+items_needing_original_publisher_updates = GenericFile.where("original_publisher" => pnb_publishers) + 
+                                           Collection.where("original_publisher" => pnb_publishers)
+
+items_needing_original_publisher_updates.each do |item|
+  item.publisher += [pnb_publishers]
```
Comment (fenekku) [inline lib/batch/batch_metadata_updates_634.rb]: Should we check that publisher contains `png_publishers` already? Is that a possibility?

File: lib/batch/batch_metadata_updates_634.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+# Updates number 2:
+#   - All items in Science in Society Collection - delete "Science in Society" 
+#   - from Original Publisher 
+
+sis_publisher = "Science in Society"
+
+def update_sis_original_publisher(id)
+  collection = Collection.find(id)
+
+  if !collection.members.empty?
```
Comment (fenekku) [inline lib/batch/batch_metadata_updates_634.rb]: do you need this if you do an `each` in Ruby?

File: lib/batch/batch_metadata_updates_634.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+sis_collection = "93c81706-47f8-49a1-88a2-3acbc971f4ed"
+update_sis_original_publisher(sis_collection)
+
+
+# Updates number 3:
+#   - All DigitalHub items - delete "Chicago, Illinois, United States" 
+#   - and delete "Chicago, IL, USA" from Location
+
+chicago_locations = ["Chicago, Illinois, United States", "Chicago, IL, USA"]
+items_needing_based_near_updates = GenericFile.where("based_near" => chicago_locations) +
```
Comment (fenekku) [inline lib/batch/batch_metadata_updates_634.rb]: What does `=>` in `"based_near" => chicago_locations` mean ? Does it mean contains "one or the other" or "both"? I am guessing it means "one or the other".

File: lib/batch/batch_metadata_updates_634.rb (hunk 1 above)
Comment (sharpattack) [reply lib/batch/batch_metadata_updates_634.rb]: So, `item.publisher` is an array of values. For some reason, using the shovel operator, `<<` was invalid. Using the `-=` or `+=` operators returns a new instance of the array, and was valid in saving.

File: lib/batch/batch_metadata_updates_634.rb (hunk 2 above)
Comment (sharpattack) [reply lib/batch/batch_metadata_updates_634.rb]: I checked and it won't save duplicate, so I think we're safe checking if it exists already.

File: lib/batch/batch_metadata_updates_634.rb (hunk 3 above)
Comment (sharpattack) [reply lib/batch/batch_metadata_updates_634.rb]: Maybe not. My first thought was to check if it had anything in the array.

File: lib/batch/batch_metadata_updates_634.rb (hunk 4 above)
Comment (sharpattack) [reply lib/batch/batch_metadata_updates_634.rb]: I did check this, and yes it returns "one or the other".

File: lib/batch/batch_metadata_updates_634.rb (hunk 1 above)
Comment (sharpattack) [reply lib/batch/batch_metadata_updates_634.rb]: But I should refactor this to `save!` to rescue instead of just returning false.

Comment (fenekku) [reply lib/batch/batch_metadata_updates_634.rb]: I was also asking if you had forgotten to commit the tests you made for this? Or maybe you ran this on your local db?

File: lib/batch/batch_metadata_updates_634.rb (hunk 2 above)
Comment (fenekku) [reply lib/batch/batch_metadata_updates_634.rb]: Interesting, `item.publisher` is a `Set` or is filtered out of redundancies on save? How did you check?

File: lib/batch/batch_metadata_updates_634.rb (hunk 3 above)
Comment (fenekku) [reply lib/batch/batch_metadata_updates_634.rb]: Let's remove that check. If `collection.members` is an empty array the `each` won't be executed.

File: lib/batch/batch_metadata_updates_634.rb (hunk 2 above)
Comment (sharpattack) [reply lib/batch/batch_metadata_updates_634.rb]: I checked this by manually testing it. I will look to see if there is a test that checks this scenario, and add one if not.

File: lib/batch/batch_metadata_updates_634.rb (hunk 3 above)
Comment (sharpattack) [reply lib/batch/batch_metadata_updates_634.rb]: okay, sounds fine to me.

File: lib/batch/batch_metadata_updates_634.rb (hunk 1 above)
Comment (sharpattack) [reply lib/batch/batch_metadata_updates_634.rb]: There already is a test that checks updating these values. It took me time to realize that none of these tests were using the `<<` operator, and that it appeared to be in a specific format that was valid for updating. That was when I tried the specific way of updating, which is using `-=` , `+=`, or just `=`.

**Pull request merged**